### DISCUSSION TITLE
How to filter Go To Definition results

### DISCUSSION THREAD

**Discussion by daniel-ames** [microsoft/vscode-cpptools#9459] [Q&A]
Hi , I have a large project that is full of many many implementations of the same functions.

For example:

```c
char* foo = GetSomeFoo(4096);
```

In my project (this is UEFI), there might be 5 or 6 implementations of `GetSomeFoo()` in as many libraries that are selected at build time by very non-standard build tools (EDK2) that are well outside the scope of any reasonable knowledge cpptools would have. Currently, cpptools will give me a list of all the different results it found, and I have to know which one is correct. This is rarely easy. I understand I can do something like this:

```json
"browse":
{
"path": [
    "${workspaceFolder}/foo/variantXYZ/Include"
]
}
```
But I have many modules that are using all the different variants depending on what module it is, all in the same project. So it's not really reasonable to expect cpptools to know which definition pertains to which code chunk I happen to be working in. I typically work in several at once because I need to make different modules work together during boot.

Therefore... I was wondering if I can make an extension that can "intercept" the results of Go To Definition and apply my own programmatic filters to present only the results that my logic dictates. I can write logic that can know which is correct based on the deep and confusing magic of UEFI. Is there a way I can filter Go To Definition results with my extension? Does cpptools export something I can read from?

**Comment by sean-mcmanus** [comment]
We currently send the go to definition results to VS Code via the LSP (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition), so I don't think it's filterable. We'd have to switch to sending the results to our TypeScript component like we do or other cases (https://github.com/microsoft/vscode-cpptools/tree/main/Extension/src/LanguageServer/Providers ), but even in that case there'd have to be some mechanism to send the results to another extension instead of to VS Code.

You could potentially use a multi-root workspace with different browse.path and files.exclude settings such that each workspace folder has different definitions available (and/or with multiple configurations for each workspace folder).

Maybe your could write a configuration provider like CMake Tools or Makefile tools, but I don't think the current configuration provider API (https://github.com/microsoft/vscode-cpptools-api) is sufficient, since it only allows setting the browse configuration and include paths, etc.

**Reply by daniel-ames** [reply]
Thanks. So to make sure I understand, I can't make my own language client and query your language server via LSP?
I don't mind making my own command that queries the language server for definitions, and then I just filter that and present my own picker.

**Reply by sean-mcmanus** [reply]
Oh, you mean sending your own LSP messages to cpptools...sure, that may work, i.e. I don't know a reason why it wouldn't.

**Reply by daniel-ames** [reply]
Yes, that's what I mean *now*, after having read more and more documentation and now have a notion of how the whole language client/server over LSP thingy works (an idea I did not have when writing my original question :) ). I've made a few simple extensions in the last week and am learning extension dev like a madman. Only a few hours ago did I learn about making a client and possibly asking cpptools' language server for definitions. Now to figure out how to do that. I've been going thru this: https://code.visualstudio.com/api/language-extensions/language-server-extension-guide
Hopefully it will teach me enough to talk to your language server! Thank you so much for your help!
